presto-rcfile/src/main/java/com/facebook/presto/rcfile/RcFileReader.java

```diff
@@ -0,0 +1,584 @@
+        }
+        else {
+            this.codec = null;
+        }
+
+        // read metadata
+        int metadataEntries = Integer.reverseBytes(input.readInt());
+        if (metadataEntries < 0) {
+            throw corrupt("Invalid metadata entry count %s in RCFile %s", metadataEntries, dataSource);
+        }
+        ImmutableMap.Builder<String, String> metadataBuilder = ImmutableMap.builder();
+        for (int i = 0; i < metadataEntries; i++) {
+            metadataBuilder.put(RcFileDecoderUtils.readLengthPrefixedString(input).toStringUtf8(), RcFileDecoderUtils.readLengthPrefixedString(input).toStringUtf8());
+        }
+        metadata = metadataBuilder.build();
+
+        // get column count from metadata
+        String columnCountString = metadata.get(COLUMN_COUNT_METADATA_KEY);
+        try {
+            columnCount = Integer.parseInt(columnCountString);
+        }
+        catch (Exception e) {
+            throw corrupt("Invalid column count %s in RCFile %s", columnCountString, dataSource);
+        }
+
+        // initialize columns
+        columns = new Column[columnCount];
+        for (int columnIndex = 0; columnIndex < columns.length; columnIndex++) {
+            columns[columnIndex] = new Column(codec);
+        }
+
+        // read sync bytes
+        syncFirst = input.readLong();
+        syncSecond = input.readLong();
+
+        // seek to first block
+        if (offset > input.position()) {
+            seekToFirstBlockIn(offset, length);
+        }
+    }
+
+    public byte getVersion()
+    {
+        return version;
+    }
+
+    public Map<String, String> getMetadata()
+    {
+        return metadata;
+    }
+
+    public int getColumnCount()
+    {
+        return columnCount;
+    }
+
+    public long getLength()
+    {
+        return length;
+    }
+
+    public long getBytesRead()
+    {
+        return dataSource.getReadBytes();
+    }
+
+    public long getRowsRead()
+    {
+        return rowsRead;
+    }
+
+    public long getReadTimeNanos()
+    {
+        return dataSource.getReadTimeNanos();
+    }
+
+    public Slice getSync()
+    {
+        Slice sync = Slices.allocate(SIZE_OF_LONG + SIZE_OF_LONG);
+        sync.setLong(0, syncFirst);
+        sync.setLong(SIZE_OF_LONG, syncSecond);
+        return sync;
+    }
+
+    @Override
+    public void close()
+            throws IOException
+    {
+        blockPosition = 0;
+        blockRowCount = 0;
+        currentChunkSize = 0;
+        input.close();
+    }
+
+    public int nextBlock()
+            throws IOException
+    {
+        blockPosition += ColumnData.MAX_SIZE;
+        currentChunkSize = min(ColumnData.MAX_SIZE, blockRowCount - blockPosition);
+
+        // do we still have rows in the current block
+        if (currentChunkSize > 0) {
+            return currentChunkSize;
+        }
+
+        // do we have a next block
+        if (input.position() >= end || input.remaining() < SIZE_OF_INT) {
```

Comment:
Even if `input.position() >= end`, the reader should not stop unless it sees the sync.